decompiler/pipeline/ssa/variable_renaming.py

```diff
@@ -334,3 +338,93 @@ def _classes_of(self, neighborhood: Iterable[Variable]) -> Iterable[Variable]:
         for neighbor in neighborhood:
             if neighbor in self._variable_classes_handler.color_class_of:
                 yield self._variable_classes_handler.color_class_of[neighbor]
+
+
+class ConditionalVariableRenamer(VariableRenamer):
+    """
+    A minimal renaming strategy, that renames the SSA-variables such that the total number of non SSA-variables is (almost) minimal.
+    Therefore, we construct color-classes by using lexicographical BFS on the interference graph. When the interference graph is chordal
+    this leads to a minimum number of possible variables.
+    """
+
+    def __init__(self, task, interference_graph: InterferenceGraph):
+        """
+        self._color_classes is a dictionary where the set of keys is the set of colors
+        and to each color we assign the set of variables of this color.
+        """
+        super().__init__(task, interference_graph.copy())
+
+        dependency_graph = dependency_graph_from_cfg(task.graph)
+
+        mapping = {}
+        for variable in self.interference_graph.nodes():
+            contracted = tuple(self._variables_contracted_to[variable])
+            for var in contracted:
+                mapping[(var,)] = contracted
+
+        # Merge nodes which need to be contracted from self._variables_contracted_to
+        dependency_graph = networkx.relabel_nodes(dependency_graph, mapping)
+
+        # counter = 0
+        # self._decorate_graph(dependency_graph, f"dep{counter}.svg")
+
+        dependency_graph.edge = dependency_graph.edges(data=True)
+        while True:
+            for u, v, _ in sorted(dependency_graph.edges(data=True), key=lambda edge: edge[2]["score"], reverse=True):
```

Comment:
You have a multigraph now. Say, edge `(x,y)` has weight 0.5, but edge `(x,z)` has now three parallel edges each with weight 0.25. Do you still want to prioritize the `(x,y)` edge over the `(x,z)` edge? This is valid, I only want to know whether you thought about this.

--> Maybe you want to use `dependency_graph.edges(data="source")`, then you can simply write `edge[2]`.

Reply:
I though about this and I think it probably makes more sens to prioritize `(x,z)` if the sum of the weight of all edges is bigger than the weight of the single edge `(x,y)`. However, I didn't know how to easily implement this. I would probably need to merge parallel edges by summing their weights. NetworkX doesn't seem to be particular helpful in these kind of things.
This is the same reason, why the algorithm is continuously re-sorting the list of all edges. Theoretically, the list of edges could always be kept sorted through graph operations, however NetworkX doesn't seem to offer any such functionality.

Reply:
I think you have two options, if we want to change this.

1. Not using a multigraph and when identifying two nodes using the `contracted_nodes` or `contracted_edges` and updating the value and name the new variable
2. Implement the merge similar to `collapse_region`  https://github.com/fkie-cad/dewolf/blob/ba2a67a998899855797ff4f160e9542d08a1e5ab/decompiler/structures/graphs/restructuring_graph/transition_cfg.py#L175C9-L175C24
3. compute the sum of all parallel edges and always get the maximum.

Reply:
I think I'm fine with not merging them. When think about it a little longer, I'm not even sure if a case like this can happen, where we have multiple edges between the same node, without the nodes interfering anyway.

Reply:
I think a have an example which may not be perfect but shows what can happen:

basic Block:
```c
b#1 = phi(0, b#2)
a = 2 * b#1 - c
b#2 = a - 5c
```
Before the renaming, we add the instructions `b#1 = 0` and b#1 = b#2` to the predecessor blocks to remove the phi-function. This leads to the following edges in the dependency graph:

`{b#1, b#2}` weight: 1
`{b#1, a}` weight: 0,125
`{a, c}` weight: 0,25
`{b#2, a}` weight: 0,25
`{b#2, c}` weight: 0,125

First, we put variables `b#1` and `b#2` into one group and merge them, leading to the following edges

`{(b#1, b#2), a}` weight: 0,125
`{(b#1,b#2), a}` weight: 0,25
`{a, c}` weight: 0,25
`{(b#1,b#2), c}` weight: 0,125

Now, the edges between `{(b#1,b#2), a}` and `{a, c}` have the same weight, but if we would consider the sum of all parallel edges, it would be clear that we want the edge `{(b#1,b#2), a}`

Reply:
Hm, yeah I didn't consider parallel edges, which point in the opposite direction. In your example this doesn't matter, as `a -> c` interferes anyway, but guess there could be case where this is not happening.
I've changed the code with f5050c0f, such that the sum of scores for parallel edges is considered instead. I'm not sure what kind of test I could create though, which verifies this behaviour.